### PR TITLE
Path sat fix

### PR DESCRIPTION
**Pull request by robdockins**
This PR fixes path satisfiability handling for SAW, and installs a new configuration option to the Online backend for enabling and disabling communication with the solver.  The new configuration option prevents the system from wasting time communicating with the solver if no queries will be issued.

### CODE REVIEW

File: crux/src/Crux/Goal.hs [hunk 1]
```diff
@@ -351,7 +352,12 @@ proveGoalsOnline sym opts ctxt explainFailure (Just gs0) =
...
        else
          prepStatus "Checking: " (countGoals gs0)
-     res <- withSolverProcess sym $ \sp ->
+
+     -- make sure online features are enabled
+     enableOpt <- getOptionSetting enableOnlineBackend (getConfiguration sym)
+     _ <- setOpt enableOpt True
+
+     res <- withSolverProcess sym (fail "Online solving not enabled!") $ \sp ->
               inNewFrame sp (go (start,end) sp goalNum gs0 nameMap)
```
Comment (atomb) [inline crux/src/Crux/Goal.hs:361]: This should never happen, but it could be useful to clarify that it's an internal error if it ever does.

**Review by atomb — APPROVED** [review]
This looks great. Thanks for fixing it!

**Comment by robdockins** [comment]
@spernsteiner, the changes in this PR break the way `concretize` works in `crux-mir`, and I'm not sure what it the best way to fix it.  Basically, the online backend will no longer actively maintain a solver connection during simulation unless path satisfiability checking is enabled, but the concretize function relies on this connection being active.

I guess there are a couple of options: have an alternate code path for doing concretization in "offline" mode, which would first collect the path conditions and to a one-shot SAT check, and use that if online solving isn't enabled for some reason.  Alternately, we could force the online solver to be enabled.

What do you think?

**Comment by spernsteiner** [comment]
It sounds like the easiest fix for now is to force-enable the online solver in crux-mir.

I don't think it makes sense to do extra work to support `concretize` in offline mode.  In practice, `concretize` is only used for panic messages, as a way of viewing (parts of) counterexamples.  There are better ways we could support that for offline mode if needed, such as adding executable counterexample support to crux-mir.

**Pull request merged**